### PR TITLE
TEN-4-summon-productionize-extended

### PR DESCRIPTION
**Pull request by southbite**

### CODE REVIEW

File: packages/happn-3/lib/providers/security-base-auth-provider.js [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    if (errorMessage) {
+      return this.accessDenied(errorMessage);
+    }
+
+    //Anything further is dealt with in the specific provider
+    return this.providerTokenLogin(credentials, previousSession, sessionId);
+  }
+
+  async providerTokenLogin() {
+    return this.accessDenied('providerTokenLogin not implemented.');
```
Comment (southbite) [inline packages/happn-3/lib/providers/security-base-auth-provider.js:176]: should be a system error

File: packages/happn-3/lib/services/utils/service.js [hunk 2]
```diff
@@ -229,6 +254,15 @@ UtilsService.prototype.replaceAll = function (str, search, replacement) {
   return str.replaceAll(search, replacement);
 };
 
+UtilsService.prototype.coerceArray = function (possibleArray) {
```
Comment (southbite) [inline packages/happn-3/lib/services/utils/service.js:257]: work out what this is doing and why

File: packages/happn-3/lib/providers/security-base-auth-provider.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    }
+    if (credentials.token) {
+      return this.tokenLogin(credentials, sessionId, request);
+    }
+
+    return this.#userCredsLogin(credentials, sessionId);
+  }
+
+  async tokenLogin(credentials, sessionId, request) {
+    let [authorized, reason] = this.#securityFacade.utils.coerceArray(
```
Comment (southbite) [inline packages/happn-3/lib/providers/security-base-auth-provider.js:123]: may not need coerceArray - promise returns an array already so assignment of output may just work without it

File: packages/happn-3/lib/providers/security-base-auth-provider.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    if (
+      previousPolicy.lockTokenToOrigin &&
+      previousSession.origin !== this.#securityFacade.system.name
+    ) {
+      errorMessage = `this token is locked to a different origin by policy`;
+    }
+
+    if (errorMessage) {
```
Comment (southbite) [inline packages/happn-3/lib/providers/security-base-auth-provider.js:167]: Early exit better - surely we just return accessDenied

File: packages/happn-3/lib/providers/security-base-auth-provider.js [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+      tokenLogin,
+      additionalInfo
+    );
+    if (session == null) {
+      throw new Error('session disconnected during login');
+    }
+    return session;
+  }
+
+  checkDisableDefaultAdminNetworkConnections(credentials, request) {
```
Comment (southbite) [inline packages/happn-3/lib/providers/security-base-auth-provider.js:227]: delete this - no longer used

File: packages/happn-3/lib/providers/security-base-auth-provider.js [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+      credentials.username === '_ADMIN' &&
+      this.#config.disableDefaultAdminNetworkConnections === true &&
+      request &&
+      request.data &&
+      request.data.info &&
+      request.data.info._local === false
+    );
+  }
+
+  #userCredsLogin(credentials, sessionId) {
```
Comment (southbite) [inline packages/happn-3/lib/providers/security-base-auth-provider.js:238]: Is it necessary to have this as a separate function?

File: packages/happn-3/lib/providers/security-happn-auth-provider.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        return this.loginFailed(credentials.username, 'Invalid credentials');
+      }
+      if (credentials.digest) return this.#digestLogin(user, credentials, sessionId);
+      const hash = await this.securityFacade.users.getPasswordHash(credentials.username);
+      // eslint-disable-next-line eqeqeq
+      if (!(await this.securityFacade.security.matchPassword(credentials.password, hash))) {
+        return this.loginFailed(credentials.username, 'Invalid credentials');
+      }
+      return this.loginOK(credentials, user, sessionId);
+    } catch (e) {
```
Comment (southbite) [inline packages/happn-3/lib/providers/security-happn-auth-provider.js:57]: try catch may not be necessary, as user existence is checked for in line 48

File: packages/happn-3/lib/providers/security-happn-auth-provider.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    if ((await this.securityFacade.security.verifyAuthenticationDigest(credentials)) !== true) {
+      return this.loginFailed(credentials.username, 'Invalid credentials', null);
+    }
+    return this.loginOK(credentials, user, sessionId);
+  }
+
+  async providerCredsLogin(credentials, sessionId) {
+    try {
+      const user = await this.securityFacade.users.getUser(credentials.username);
```
Comment (southbite) [inline packages/happn-3/lib/providers/security-happn-auth-provider.js:46]: check if we are not getting the hash here - and just use what we have

File: packages/happn-3/test/unit/security/replication.js [hunk 9]
```diff
@@ -1,5 +1,5 @@
 const async = require('async');
-describe(
+xdescribe(
```
Comment (southbite) [inline packages/happn-3/test/unit/security/replication.js:2]: Get these tests done

File: packages/happner-2/lib/system/shared/mesh-client.js [hunk 10]
```diff
@@ -269,6 +275,7 @@
           Logger: opts.log,
           reconnect: opts.reconnect,
           socket: opts.socket,
+          loginRetry: opts.loginRetry,
```
Comment (southbite) [inline packages/happner-2/lib/system/shared/mesh-client.js:278]: Why is this here?

**Review by southbite — COMMENTED** [review]
Changes required

File: packages/happner-2/lib/system/shared/mesh-client.js (hunk 10 above)
Comment (southbite) [reply packages/happner-2/lib/system/shared/mesh-client.js:278]: was not finding its way to happn client,whilst testing summon provider

File: packages/happn-3/lib/providers/security-base-auth-provider.js (hunk 3 above)
Comment (southbite) [reply packages/happn-3/lib/providers/security-base-auth-provider.js:127]: leaving it in for now

File: packages/happn-3/test/unit/security/replication.js (hunk 9 above)
Comment (southbite) [reply packages/happn-3/test/unit/security/replication.js:2]: deleted, this is extensively unit tested already

**Review by leebow — APPROVED** [review]
